### PR TITLE
Stable calculation of quaternion from transformation matrix

### LINKED ISSUE #3588: Quaternions.from_T returns error for valid matrix
In some cases, `Modelica.Mechanics.MultiBody.Frames.Quaternions.from_T` returns error even for a valid matrix input, see the following code:

```
model CheckQfromT
  import Modelica.Mechanics.MultiBody.Frames.Quaternions;

  parameter Real T[:,:] = {{0,1,0},{1,0,0},{0,0,-1}} "Valid transformation matrix";
  parameter Real E[:,:] = T*transpose(T) "To check orthogonality of T";
  parameter Real D = Modelica.Math.Matrices.det(T) "Determinant (shall be = 1)";
  parameter Real t = Modelica.Math.Matrices.trace(T) "Indentify a problem if and only if trace = -1";
  parameter Quaternions.Orientation Q = Quaternions.from_T(T) "Quaternions to be calculated";
equation 
end CheckQfromT;

```

A possible solution is given here: https://d3cw3dd2w32x2b.cloudfront.net/wp-content/uploads/2015/01/matrix-to-quat.pdf

### PR DESCRIPTION
**Pull request by tobolar**
Close #3588
(It could be possible to apply this to older version(s) as well.)

### CODE REVIEW

File: Modelica/Mechanics/MultiBody/Frames/Quaternions/from_T.mo [hunk 1]
```diff
@@ -10,63 +10,42 @@ function from_T
...
+  else
+    if T[1, 1] < -T[2, 2] then
+      t := 1 - T[1, 1] - T[2, 2] + T[3, 3];
+      Q := {T[3, 1] + T[1, 3], T[2, 3] + T[3, 2], t, T[1, 2] - T[2, 1]};
+    else
+      t := 1 + T[1, 1] + T[2, 2] + T[3, 3];
+      Q := {T[2, 3] - T[3, 2], T[3, 1] - T[1, 3], T[1, 2] - T[2, 1], t};
+    end if;
+  end if;
+  Q := Q .* 0.5/sqrt(t);
```
Comment (beutlich) [inline Modelica/Mechanics/MultiBody/Frames/Quaternions/from_T.mo:36]:
```suggestion
  Q := Q*0.5/sqrt(t);
```
The component-wise multiplication is not relevant when multiplying by a scalar and made me wondering first.

**Review by beutlich — COMMENTED** [review]
Two more thoughts:

* Would be good to have your test model of #3588 in ModelicaTest as simple unit test.
* Maybe you could add a proper Documentation.revision annotation where you summarize the improvement and also link to the utilized source.

**Comment by tobolar** [comment]
@beutlich Done.

**Comment by beutlich** [comment]
> 
> 
> @beutlich Done.

I cleaned up the commit history.

**Comment by christiankral** [comment]
I am afraid I cannot asses this issue.

**Review by beutlich — CHANGES_REQUESTED** [review]

**Review by HansOlsson — APPROVED** [review]
Looks good. I also like that the statement c1>=0, c2>=0 ... was removed; as it seems false.

**Review by MartinOtter — APPROVED** [review]

**Comment by beutlich** [comment]
I rebased the PR branch to include the latest CI checks.

**Pull request merged**